### DISCUSSION TITLE
How long does collection recovery take

### DISCUSSION THREAD

**Discussion by pycui** [milvus-io/milvus#35757] [Q&A and General discussion] [answered]
Hi - I'm testing a standalone deployment with about 49M rows right now. The standalone service was evicted by k8s and restarted. 
Now when I try to do search: I always get this `pymilvus.exceptions.MilvusException: <MilvusException: (code=106, message=failed to search: loaded collection do not found any channel in target, may be in recovery: collection on recovering[collection=452125728648336710])>`
It's about 77 minutes since restarting. Wondering what's the expected recovery time here? Is this normal? thanks

**Comment by yhmo** [comment, marked as answer]
When the service restarts, it tries to load the collection into the memory. The error "... may be in recovery" indicates it failed to load the collection, perhaps due to resource limit(insufficient memory maybe).
For further investigation, we need the full log when restarting. You can use `kubectl logs <pod-name>` to fetch the logs.

**Reply by pycui** [reply]
Thank you for the reply. 
Here is log for milvus and minio after restarting. I think it's already truncated by k8s due to too long
[milvus.log](https://github.com/user-attachments/files/16777824/milvus.log)
is the 
[minio.txt](https://github.com/user-attachments/files/16777825/minio.txt)

Given your response, I'm wondering if the data is already lost / corrupted? I'm using the official helm chart https://github.com/zilliztech/milvus-helm/tree/master/charts/milvus with only one change: since my k8s cluster doesn't have storageclass, I use standalone.persistence.enabled = false, and change the milvus-data-disk to a persistent hostPath

```
Volumes:
  tools:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
  milvus-config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      my-release-milvus
    Optional:  false
  milvus-data-disk:
    Type:          HostPath (bare host directory volume)
    Path:          /data/milvus
    HostPathType:  DirectoryOrCreate
  disk:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
```
Similarly for minio
```
        - name: export
          hostPath:
            path: /data/milvus
            type: DirectoryOrCreate
```

Can this cause problem? 

In general, what's your suggestion for persistence without using PVC (e.g. hostPath)? Thank you!!

**Reply by yhmo** [reply]
There are a lot of errors like this: 
`[2024/08/28 07:10:33.276 +00:00] [WARN] [segments/segment.go:1024] ["CStatus returns err"] [traceID=c4d1a810e564c35a27391788dbd94974] [collectionID=452125728646630728] [partitionID=452125728646630729] [segmentID=452125728646830759] [fieldID=101] [error="Error in GetObjectSize[errcode:404, exception:, errmessage:No response body., params:params, bucket=milvus-bucket, object=file/insert_log/452125728646630728/452125728646630729/452125728646830759/101/452125728646833872]"] [extra="LoadFieldData failed"]`

Two dependencies for milvus:
1. MinIO/S3 is the major data storage of Milvus, all the original data and index data is stored here as blobs.
2. Etcd stores the system metadata of Milvus, including schema of collections, runtime address of each component, paths of data/index of each segment, etc.

The metadata in Etcd must be consistent with blobs in MinIO. When milvus is loading a collection, it collects the paths from Etcd, and finds the blobs in MinIO according to the paths. If the persistence configuration is changed, the metadata in Etcd and blobs in MinIO are no longer consistent, then you will get the errors.

**Reply by pycui** [reply]
Given the restart is caused by k8s, the persistence configuration shouldn't be changed (also verified by checking the pod describe). I'm wondering if using hostPath doesn't work actually. Can you shed some light?

**Reply by LoveEachDay** [reply]
@pycui After eviction, there's no guarantee that the Milvus and MinIO pods will be scheduled back to their original worker nodes, especially when using hostPath volumes. If these pods are scheduled to new worker nodes, the data may be lost, resulting in a failure to recover your Milvus cluster.

**Reply by pycui** [reply]
Thanks for the reply, I should be more clear, the hostPath we use is a distributed filesystem that is available cluster-wise. So the hostPath should still be available. 
A separate question is, in general, if Milvus and MinIO are restarted but not the etcd (and assuming volume is OK), is this situation safe to recover?

**Reply by yhmo** [reply]
> Thanks for the reply, I should be more clear, the hostPath we use is a distributed filesystem that is available cluster-wise. So the hostPath should still be available. A separate question is, in general, if Milvus and MinIO are restarted but not the etcd (and assuming volume is OK), is this situation safe to recover?

If the volume is ok, and minio and etcd are healthy, milvus can recover successfully. 

Etcd is a key dependency for milvus. If etcd is unhealthy, milvus is not able to work and terminates itself after a few minutes.
If MinIO is down, milvus can partly work(search, insert still work), but it is not able to persist any data. For standalone there is a build-in WAL component that ensure the data is safe. For cluster the Pulsar/Kafka is like a WAL component. Once the MinIO is back to work, milvus can fetch the data from WAL component to recover.

**Reply by pycui** [reply]
Thanks for the reply. I checked the configs and I think this may be because the `export` volume of MinIO is still EmptyDir even though we changed to hostPath in deployment.yaml. Can you comment if `export` volume is indeed important for data persistence?
